NotebookPropertiesDialog Component To Spec For Review

bedwardly-down opened the pull request:
<!--

Please prefix the title with the platform you are targetting:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

For example: "Desktop: Added new setting to change font", or "Mobile: Fixed config screen error"

PLEASE READ THE GUIDE FIRST: https://github.com/laurent22/joplin/blob/master/CONTRIBUTING.md

-->

I spent the last two days taking your feedback and working on learning how React Hooks work, and also reevaluating my code, because, let's be frank, my previous submission didn't work a quarter as good as this so far. That's because I went off spec and tried to take the easy way out and pretty much copy by hand a different component to see how the code flows. This time I took my time, went through the spec and built the NotebookPropertiesDialog component in chunks.

## Tasks Completed

- Add a new icon property to the Folder object - it will be a string that will take any of the Fork Awesome icon names.
- Make sure Fork Awesome is loaded in the app (it should already be) so that its icons can be displayed
- When right-clicking a notebook, change "Rename" to "Properties"
- Change the Rename dialog to a more generic Notebook Properties dialog
- Add the option to set the icon there. Initially it can be a simple text input where the user inputs the name of the icon directly. We can link to the Fork Awesome page to give the list of possible icons.

## Tasks mostly completed with bugs

- When the user click Save, save the icon to the database
- In the sidebar, display the icon next to the notebook name.

The final two commits were my attempts at getting these two parts to function properly, but I am a bit at a loss for why neither one are working properly and welcome feedback.

tessus commented:
@bedwardly-down please install the pre-commit hook. There are a lot of eslint issues.

bedwardly-down commented:
@tessus let me reinstall it because it was installed. I wonder if there may be a conflict going on with my editor.

bedwardly-down commented:
So, disabling the linters in my editors and then submitting an irrelevant change to allow for a commit doesn't solve the problem. Here's what my terminal shows:

> [I] ⋊> ~/S/g/joplin on NPD-Redux ⨯ git commit                             14:48:55
husky > pre-commit (node v13.7.0)
  ↓ Stashing changes... [skipped]
    → No partially staged files found...
  ✔ Running tasks...
[NPD-Redux dc38a1df] Linter Conflict Fix?
 4 files changed, 4 insertions(+), 4 deletions(-)

tessus commented:
Hmm, I have no idea what's going on here.

tessus commented:
Let me try to pull your PR, revert the last commit and let's see what happens...

bedwardly-down commented:
Thanks for checking into this, @tessus . Worse case scenario, I retype it all with a linter free development environment now that I learned what linters actually were used for. I just don't use them enough.

tessus commented:
Fixed it. The file that is created from the typescrpt file must not be checked with the linter, thus I've added it to the .eslintignore. See the commit I pushed. To get it locally, you will have to pull it.

bedwardly-down commented:
It won't be until at least 11 pm CST tonight before I can pull it down, but thanks for the help again. ☺️ 

It's time to go back to work. Grrr. Ha

bedwardly-down commented:
What I've noticed so far on my end is that the linter still hasn't affected NotebookPropertiesDialog.tsx, so the eslint problems are still there. Like I said, there's no issue for me rewriting the code again on a clean branch to make it fit the style you are needing for this project. I would like some feedback on what I can improve on or do to possibly fix the bugs I am running into.

The save bug: the console shows that the Icon information is being saved and the app starts syncing only to have the Icon be undefined when I open the properties dialog again. It seems to be overwriting it somehow on close. I do have some ideas why this may be, but the fixes I would want to implement would be off spec and could possibly break other things.

The icon bug: until I get the previous bug fixed, I have little to no way to test this out to see what's what with it.

tessus commented:
Forget the linter issues. There aren't any anymore. Don't you see the green checkmark next to your PR?
Travis CI errored out due to a generated file being lintered. Didn't I mention this in my last comment?

With regards to the bugs, maybe @laurent22 has an idea what's going on.

bedwardly-down commented:
Thanks @tessus. i also just may have realized why the save bug is happening. I'll test out a theory when i get home from work tonight.

bedwardly-down commented:
My work is contract work and I made enough today that tomorrow, I can take off. I was on a programming server discord asking about the two issues I'm having with the code and apparently both are very basic fixes that require me studying up on some Javascript fundamentals. I'll spend tomorrow working on that and then see if I can get my main idea to work properly and finish this task.

bedwardly-down commented:
I'm still having problems getting the icon to save. It seems to save the variable in memory only to not save to the database.

laurent22 commented:
You need to call `Folder.save({ id: FOLDER_ID, icon: FOLDER_ICON })` somewhere. Do you do this, and what's the result?

bedwardly-down commented:
I don't do that. I followed what the rename structure was in the renameFolder Command that I replaced. I can definitely try that. Thanks.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+export default function NotebookPropertiesDialog(props: NotebookPropertiesDialogProps) {
+	console.info('Render NotebookPropertiesDialog');
+
+	const [notebook, setNotebook] = useState<any>({});
```
Please use the term "folder" internally. I know it's inconsistent that it's called "Notebook" in the UI and "folder" internally, but at least we can try to keep it consistent internally. So nearly all the time it should be called "folder", except in strings in the UI.

laurent22 replied:
Please clearly name what representation of a folder this is. In this case, it's the form own representation (it only needs to have an ID, title and icon) so name it "formFolder" so that we know. When it's just "folder" it means it's the representation as it is in the database.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+		<div style={theme.dialogModalLayer}>
+			<div style={rootStyle}>
+				<div style={theme.dialogTitle}>{_('Notebook Properties')}</div>
+				<label style={theme.textStyle}>{_('Icon: ')}</label>
```
It should also be possible to change the title of the notebook in this dialog, since it replaces the Rename dialog.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+				icon: targetNotebook.icon,
+			});
+			setIcon(nb.icon);
+			setLoaded(true);
```
If you have a `formFolder` state variable, you don't need to also have an icon state variable, because there's already `formFolder.icon`.

bedwardly-down replied:
In my current test branch, this has been removed for that reason.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+				icon: targetNotebook.icon,
+			});
+			setIcon(nb.icon);
+			setLoaded(true);
```
`setLoaded` should be called from where the data is actually loaded.

bedwardly-down replied:
Removed due to having no actual value here for what's happening. I can definitely add back if absolutely needed but it was a fragment of some tests I was running to see why icon saving and loading wasn't working.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+		fetchNotebook();
+	}, [props.folderId]);
+
+	const formNotebook = (targetNotebook:any) => {
```
Not sure about the purpose of this function. You've already defined a `setNotebook` (should be called `setFormFolder`) so this seems redundant. I think you can do all the loading and setting state in the useEffect above.

bedwardly-down replied:
I figured that out in my current test branch. Just need to work out the kinks before pushing it.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+	const saveNotebook = async () => {
+		console.info('Saving Notebook Now');
+		if (!notebook) {
+			console.error('Notebook saving failed in the NotebookPropertiesDialog');
```
Can't happen. If it doesn't it's an exceptional error so an exception should be thrown - `throw new Error(msg)`

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+			console.error('Notebook saving failed in the NotebookPropertiesDialog');
+		} else {
+			const nb = await Folder.save(notebook, {userSideValidation: true});
+			setNotebook(nb);
```
The folder is not modified here so doesn't need to be set.

bedwardly-down commented:
> You need to call `Folder.save({ id: FOLDER_ID, icon: FOLDER_ICON })` somewhere. Do you do this, and what's the result?

> 2020-02-04 09:37:06: "Scheduling sync operation..."
2020-02-04 09:37:16: "SearchEngine: Updating FTS table..."
2020-02-04 09:37:16: "SearchEngine: Updated FTS table in 4ms. Inserted: 0. Deleted: 0"

Now this is running from the NotebookPropertiesDialog component. I think it's closing out before it's actually saving, but I could be wrong.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,66 @@
+export default function NotebookPropertiesDialog(props: NotebookPropertiesDialogProps) {
+	console.info('Render NotebookPropertiesDialog');
+
+	const [folder, formFolder] = useState<any>({});
```
Not sure if it's a typo but it should be `const [formFolder, setFormFolder]`

bedwardly-down replied:
Not a typo, just a misunderstanding. form is usually a verb, so I figured you were requesting set, another verb, to be changed to that. It's not quite clear elsewhere in the code that form is being used as an adjective here. I can definitely fix that.

laurent22 replied:
Sorry I didn't mean as a verb. "form" in this context means the form where you enter data. So this `formFolder` will have the current state of the form.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,66 @@
+
+	const saveIcon = async () => {
+		await Folder.save({id: folder.id, icon: folder.icon, userValidation: true});
+	};
```
Thanks, it looks quite good now but we don't need two separate save handlers. When the user press the Save button on the dialog, it should save both icon and title. And userValidation is the second argument.

     await Folder.save({id: folder.id, icon: folder.icon, title: folder.title }, {userValidation: true});

bedwardly-down replied:
I can definitely do that.

bedwardly-down replied:
Replaced Save Button with the fa-save icon to conform with appearance of other elements in Joplin.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,66 @@
+		<div style={theme.dialogModalLayer}>
+			<div style={rootStyle}>
+				<div style={theme.dialogTitle}>{_('Notebook Properties')}</div>
+				<label style={theme.textStyle}>{_('Name:   ')}</label>
```
Should use styling to align elements, especially since this is a string to be translated and the spacing would be different from one language to another.

bedwardly-down replied:
I will need to study a tiny bit on this. I do have work this afternoon and do need to run some errands beforehand in a bit.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,66 @@
+	useEffect(() => {
+		async function fetchFolder() {
+			if (!props.folderId) {
+				formFolder({folder: null});
```
Better to throw an exception as it won't work with an empty folder.

bedwardly-down replied:
Can definitely do.

laurent22 commented on ElectronClient/app/gui/NotebookPropertiesDialog.tsx:
```diff
@@ -0,0 +1,66 @@
+	onClose: Function,
+}
+
+export default function NotebookPropertiesDialog(props: NotebookPropertiesDialogProps) {
```
Also needs to be renamed to FolderPropertiesDialog and FolderPropertiesDialogProps.

bedwardly-down replied:
I made a typo that fixed this. git revert here I come. :D

bedwardly-down commented:
When I get home from work, I will definitely resolve the rest of the issues here. Thanks again for reviewing.

laurent22 commented on ElectronClient/app/gui/MainScreen.jsx:
```diff
@@ -137,15 +144,15 @@ class MainScreenComponent extends React.Component {
 			const noteTags = tags
 				.map(a => {
-					return { value: a.id, label: a.title };
+					return {value: a.id, label: a.title};
```
I don't understand why this is being reformatted, or why it's not picked by the linter as the preferred syntax would be the one that's been deleted.

bedwardly-down replied:
I've disabled / uninstalled all of the linters in my editor that i know of. I'm not even touching any parts outside of what I'm currently working on.

laurent22 replied:
That was my mistake actually, I thought there was a rule for this but there wasn't. So I've just added one, as well as for array spacing.

bedwardly-down replied:
Ale is a universal linter that works for almost all languages. Of course, i did notice that the eslint script is only using the default linter settings and can be custom configured using the `--config` or `--conf` flags. Maybe the defaults changed too?

bedwardly-down replied:
Despite this, I'm always up for trimming my plugins and configs at times. I need to nuke my neovim config a bit. I have quite a bit of junk in it and coc.nvim is handling most of my coding needs. VsCode is a hog on my system but i love Intellisense and some of its plugins. Coc adds full support of both.

laurent22 commented:
> EDIT: It's not closing out with any issues from that. It's saving title fine but not saving or loading icon even after setting a default for it.

Right, I should have mentioned that. It won't save the icon because you need first to add a migration to modify the database and add an "icon" field. Have a look in joplin-database to see how it's done.

Column should be not null with default "".

laurent22 commented:
By the way it's a good idea to open the sqlite database and see for yourself what is saved or not. The table you need here is simply called "folder"

bedwardly-down commented:
> By the way it's a good idea to open the sqlite database and see for yourself what is saved or not. The table you need here is simply called "folder"

Will do. I'm learning more than i thought i would here. Ha

bedwardly-down commented:
When I finish this to your liking, @laurent22, so that the commits are much cleaner, I can definitely create patches and patch a clean branch from the latest commits and then send a new pull request that is committed to match the specs on this task. That's how I did most of the commits until I hit the trouble point I reached.

bedwardly-down commented:
Also, I tried pulling the new linter commit using cherry pick and it pulled a massive amount of other commits that were conflicting not with my code but other parts. That's part of why you see the Translation commit and why the new linter rule isn't added yet. Definitely why I want to finish, create patches and commit just my changes on a clean branch.

tessus commented:
Usually you do a `git rebase`, if you want to pull changes from master into this branch. You'd then have to force push your branch, but since it is still in development, this would be fine.

You can also squash your commits in this branch into one commit (all but the language update commit) and then rebase.

Whatever you do, please make a backup, if you don't know git very well. Otherwise your changes might be lost forever.

bedwardly-down commented:
I'll check that out. I've never used squash or rebase. I mainly just backup my code to a directory outside of my repo, keep that branch alive until I am certain that my new, fresh branch is still good, then patch in the changes from my working files to the new branch, then, once everything is good, I remove the old branch.

laurent22 commented:
@bedwardly-down, please ping me when this is finished and ready for review.

bedwardly-down commented:
@tessus thanks for the tip. I will definitely do just that.

@laurent22, everything is working like it should so far, although, I'm going to rebuild my database and check the code for errors. If you have any comments on any of the styling or other things, please post them now. If everything is good, I'll start squashing my commits and getting everything ready to finish this puppy up.

bedwardly-down commented:
@tessus, I tried squashing the commits and it's just messy. Grrr

bedwardly-down commented:
@laurent22, this task is done and up for your review. I am not touching this anymore until it's safe to do so and really need to brush up on git. I thought I knew how to use it and don't want to botch this anymore.

tessus commented:
@bedwardly-down yep, I checked. It's because you messed up earlier already. I'm sorry, I can't fix that. Not unless I invest a lot of time in conflict solving and so on.

bedwardly-down commented:
Done. @laurent22

bedwardly-down commented:
New branch made with cleaned up commits and no bugs so far. It is called NPD-Redux-Clean and a new pull request has been made for it.

Rishgod commented:
Hey, @bedwardly-down what the status for this, maybe I can help you out if you want?

bedwardly-down commented:
> Hey, @bedwardly-down what the status for this, maybe I can help you out if you want?

This issue is for grabs. I made it pretty far but had to take a step back due to personal reasons. If you want to see what I've done and think you can do the task justice, my branch should still be available. 

`git clone https://github.com/bedwardly-down/joplin -b NBD-Redux-Clean`
